### PR TITLE
Added new TC MutipleBackupLocationWithSameEndpoint

### PR DESCRIPTION
**Pull request by ak-px**
**What this PR does / why we need it**:
Added new TC MutipleBackupLocationWithSameEndpoint
[Multiple backup location with same/different nfs targets](https://portworx.testrail.net/index.php?/cases/view/84902)
Step 2 is not add in this TC , as we are covering the scenario already in CloudSnapsSafeWhenBackupLocationDeleteTest.

**Which issue(s) this PR fixes** (optional)
Closes #PA-1494

**Special notes for your reviewer**:
S3:
https://jenkins.pwx.dev.purestorage.com/blue/organizations/jenkins/portworx-backup%2Fsystem-tests%2Fbyoc%2Fpx-backup-on-demand-system-test-byoc/detail/px-backup-on-demand-system-test-byoc/2777/
https://aetos.pwx.purestorage.com/resultSet/testSetID/358946 

NFS :
In Progress

### CODE REVIEW

File: tests/backup/backup_scale_test.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+		DestroyApps(scheduledAppContexts, opts)
+		log.InfoD("Deleting the px-backup objects")
+		CleanupCloudSettingsAndClusters(backupLocationMap, cloudCredName, cloudCredUID, ctx)
+		log.InfoD("Switching context to destination cluster for clean up")
+		err = SetDestinationKubeConfig()
+		log.FailOnError(err, "Unable to switch context to destination cluster [%s]", destinationClusterName)
+		DestroyApps(scheduledAppContexts, opts)
+		log.InfoD("Switching back context to Source cluster")
+		err = SetSourceKubeConfig()
+		log.FailOnError(err, "Unable to switch context to source cluster [%s]", SourceClusterName)
```
Comment (mkoppal-px) [inline tests/backup/backup_scale_test.go:209]: Put this in a defer block

File: tests/common.go [hunk 2]
```diff
@@ -9254,3 +9257,25 @@ func GetGkeSecret() (string, error) {
...
+		Name:  cloudCredentialName,
+		Uid:   cloudCredentialUid,
+	}
+	f := func() (interface{}, bool, error) {
+		inspectRsp, err := Inst().Backup.InspectCloudCredential(ctx, cloudCredentialrInspectReq)
+		if err == nil {
+			return nil, false, nil
+		} else {
+			return nil, true, fmt.Errorf("Failed to inspect cloudCredential [%v] with response [%v] . Waiting to become complete", cloudCredentialName, inspectRsp)
+		}
```
[the remark this answers is not in the document]
Comment (ak-px) [reply tests/common.go]: Done.

File: tests/backup/backup_scale_test.go (hunk 1 above)
Comment (ak-px) [reply tests/backup/backup_scale_test.go]: Done

File: tests/backup/backup_helper.go [hunk 3]
```diff
@@ -2168,25 +2168,6 @@ func IsBackupLocationPresent(bkpLocation string, ctx context.Context, orgID stri
 	return false, nil
 }
 
-// IsCloudCredPresent checks whether the Cloud Cred is present or not
```
Comment (kshithijiyer-px) [inline tests/backup/backup_helper.go:2171]: Why do you want to move this?

Comment (ak-px) [reply tests/backup/backup_helper.go:2171]: Need to use it in CreateS3BackupLocation in common.go for checking the cloudcred status.

Comment (kshithijiyer-px) [reply tests/backup/backup_helper.go:2171]: Okay

File: tests/backup/backup_scale_test.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+				scheduledAppContexts = append(scheduledAppContexts, appCtx)
+				namespace := GetAppNamespace(appCtx, taskName)
+				bkpNamespaces = append(bkpNamespaces, namespace)
+			}
+		}
+	})
+
+	It("Create Backup and Restore for Mutiple backup location added using same endpoint", func() {
+		Step("Validate applications", func() {
+			log.InfoD("Validating applications")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_scale_test.go:57]: Can we move this below ValidateApplications?

Comment (ak-px) [reply tests/backup/backup_scale_test.go]: Done

File: tests/backup/backup_scale_test.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+					wg.Add(1)
+					go createBackup(backupName, namespace, index)
+					// wait for max concurrent backup limit of 4 to be completed
+					if index%4 == 0 {
+						wg.Wait()
+					}
+				}
+			}
+			wg.Wait()
+		})
```
Comment (kphalgun-px) [inline tests/backup/backup_scale_test.go:123]: How about making using of semaphore?

```
semaphore := make(chan int, 4)
for _, namespace := range bkpNamespaces {
	for index := 0; index < numberOfBackups; index++ {
		time.Sleep(timeBetweenConsecutiveBackups)
		backupName := fmt.Sprintf("%s-%s-%s", BackupNamePrefix, backupLocationNameMap[index], RandomString(4))
		userBackupMap[index] = make(map[string]string)
		userBackupMap[index][backupName] = namespace
		wg.Add(1)
		semaphore <- 0
		go func(backupName string, namespace string, index int) {
			defer wg.Done()
			defer func() {
				<-semaphore
			}()
			createBackup(backupName, namespace, index)
		}(backupName, namespace, index)
	}
}
wg.Wait()
```

I have not tested it.

File: tests/backup/backup_scale_test.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+					go createRestore(backupName, restoreName, namespace)
+				}
+			}
+			wg.Wait()
+		})
+		Step("Delete all Backup locations from px-admin", func() {
+			log.InfoD("Delete Backup locations from px-admin")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "failed to fetch ctx for admin")
+			AllBackupLocationMap, err := getAllBackupLocations(ctx)
```
Comment (kphalgun-px) [inline tests/backup/backup_scale_test.go:150]: Can we make use of backupLocationMap?

File: tests/backup/backup_scale_test.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+		log.InfoD("Deleting the deployed apps after the testcase")
+		opts := make(map[string]bool)
+		opts[SkipClusterScopedObjects] = true
+		DestroyApps(scheduledAppContexts, opts)
+		log.InfoD("Deleting the px-backup objects")
+		CleanupCloudSettingsAndClusters(backupLocationMap, cloudCredName, cloudCredUID, ctx)
+		log.InfoD("Switching context to destination cluster for clean up")
+		err = SetDestinationKubeConfig()
+		log.FailOnError(err, "Unable to switch context to destination cluster [%s]", destinationClusterName)
+		DestroyApps(scheduledAppContexts, opts)
```
Comment (kphalgun-px) [inline tests/backup/backup_scale_test.go:209]: I think this would not work because of custom namespace mapping.

File: tests/common.go [hunk 8]
```diff
@@ -4276,6 +4289,20 @@ func CreateS3BackupLocation(name string, uid, cloudCred string, cloudCredUID str
...
+		if err != nil {
+			return "", true, fmt.Errorf("cloud cred %s present with error %v", cloudCred, err)
+		}
+		if status {
+			return "", true, nil
+		}
+		return "", false, nil
+	}
+	_, err := task.DoRetryWithTimeout(cloudCredStatus, defaultTimeout, defaultRetryInterval)
+	Inst().Dash.VerifySafely(err, nil, fmt.Sprintf("Verfiying cloud cred %s is present", cloudCred))
```
Comment (kphalgun-px) [inline tests/common.go:4304]: I feel this should be done while creating cloud credential. @mkoppal-px

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: tests/backup/backup_scale_test.go (hunk 5 above)
Comment (ak-px) [reply tests/backup/backup_scale_test.go]: Added  semaphore and it working as expected, Thanks for the suggestion @kphalgun-px 
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/2988/

File: tests/backup/backup_scale_test.go (hunk 7 above)
Comment (ak-px) [reply tests/backup/backup_scale_test.go]: Right, removed the code for now.

File: tests/backup/backup_scale_test.go (hunk 6 above)
Comment (ak-px) [reply tests/backup/backup_scale_test.go:153]: Done

File: tests/common.go (hunk 8 above)
Comment (ak-px) [reply tests/common.go]: I also think that would be better instead of checking status for all type of location.  Moved the code to CreateCloudCredential.

**Comment by mkoppal-px** [comment]
Can you trigger a run for all tests on Vanilla since we have altered the core function of creating backup location?

**Comment by ak-px** [comment]
@mkoppal-px 
aetos lnk all test run on vanilla.
https://aetos.pwx.purestorage.com/resultSet/testSetID/383637 
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/basic-system-test/job/px/job/s3/job/px-backup-system-test-vanilla/491/
The few failure in the runs are in cleanup  due to dependency issue, No failure seen creating backup location/ cloud credentials.

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**